File: LeetCode/328.odd-even-linked-list.cpp
/*
 * @lc app=leetcode id=328 lang=cpp
 *
 * [328] Odd Even Linked List
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class SolutionA
{
    // Remember to cut off the link between current element and the rest of the list
public:
    ListNode *oddEvenList(ListNode *head)
    {
        ListNode odd(0), even(0);
        ListNode *oddEnd = &odd, *evenEnd = &even;
        int i = 1;
        while (head != nullptr)
        {
            ListNode *ele = head;
            head = head->next;
            ele->next = nullptr;
            if (i & 1)
            {
                oddEnd->next = ele;
                oddEnd = ele;
                i = 0;
            }
            else
            {
                evenEnd->next = ele;
                evenEnd = ele;
                i = 1;
            }
        }
        oddEnd->next = even.next;
        return odd.next;
    }
};

class SolutionB
{
public:
    ListNode *oddEvenList(ListNode *head)
    {
        if (head == nullptr)
            return nullptr;
        ListNode *oddEnd = head, *evenEnd = head->next;
        while (evenEnd && evenEnd->next)
        {
            ListNode *ele = evenEnd->next;

            // Skip even element.
            evenEnd->next = ele->next;
            evenEnd = ele->next;

            // Insert odd element behind the end of odd elements.
            ele->next = oddEnd->next;
            oddEnd->next = ele;
            oddEnd = ele;
        }
        return head;
    }
};

class Solution
{
public:
    ListNode *oddEvenList(ListNode *head)
    {
        if (head == nullptr)
            return nullptr;
            
        ListNode *oddEnd = head, *evenEnd = head->next, *evenHead = head->next;
        while (evenEnd && evenEnd->next)
        {
            ListNode *ele = evenEnd->next;

            oddEnd->next = ele;
            oddEnd = oddEnd->next;
            evenEnd->next = ele->next;
            evenEnd = evenEnd->next;
        }
        oddEnd->next = evenHead;
        return head;
    }
};

File: LeetCode/java/328.odd-even-linked-list.java
/*
 * @lc app=leetcode id=328 lang=java
 *
 * [328] Odd Even Linked List
 *
 * https://leetcode.com/problems/odd-even-linked-list/description/
 *
 * algorithms
 * Medium (55.38%)
 * Likes:    2714
 * Dislikes: 329
 * Total Accepted:    341.6K
 * Total Submissions: 600.3K
 * Testcase Example:  '[1,2,3,4,5]'
 *
 * Given a singly linked list, group all odd nodes together followed by the
 * even nodes. Please note here we are talking about the node number and not
 * the value in the nodes.
 * 
 * You should try to do it in place. The program should run in O(1) space
 * complexity and O(nodes) time complexity.
 * 
 * Example 1:
 * 
 * 
 * Input: 1->2->3->4->5->NULL
 * Output: 1->3->5->2->4->NULL
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: 2->1->3->5->6->4->7->NULL
 * Output: 2->3->6->7->1->5->4->NULL
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The relative order inside both the even and odd groups should remain as it
 * was in the input.
 * The first node is considered odd, the second node even and so on ...
 * The length of the linked list is between [0, 10^4].
 * 
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode oddTail = head;
        ListNode cur = head.next;
        while (cur != null && cur.next != null) {
            ListNode ele = cur.next;
            cur.next = ele.next;
            ele.next = oddTail.next;
            oddTail.next = ele;
            oddTail = ele;
            cur = cur.next;
        }
        return head;
    }
}
// @lc code=end
